Accept TOML format in rust-toolchain

ebroto opened the pull request:
Closes #2350

ebroto committed:
Accept TOML format in rust-toolchain

ebroto commented:
Not sure about the macOS CI failure as I can't see the logs :thinking:

rbtcollins commented:
Builder probably flaked, just run it again

On Fri, 24 Jul 2020, 10:32 ebroto, <notifications@github.com> wrote:

> Not sure about the macOS CI failure as I can't see the logs 🤔
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rustup/pull/2438#issuecomment-663264420>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADZ7XSFKR7ASWFNOICTL43R5C3BPANCNFSM4PGEQPKA>
> .
>

kinnison commented on README.md:
````diff
@@ -404,9 +404,21 @@ source repository. This is most often the case for nightly-only
+targets = [ "wasm32-unknown-unknown", "thumbv2-none-eabi" ]
+```
+
+If the TOML format is used, the `[toolchain]` section and the `channel` entry
````
I wonder if we need `channel` to be mandatory -- people may want to use this new more flexible format just to add a target such as `wasm32-unknown-unknown` without stipulating a particular channel.

kinnison commented on src/cli/common.rs:
```diff
@@ -480,12 +480,12 @@ pub fn list_toolchains(cfg: &Cfg, verbose: bool) -> Result<utils::ExitCode> {
-        };
+        let ovr_toolchain_name =
+            if let Ok(Some((toolchain, _override_cfg, _reason))) = cfg.find_override(&cwd) {
+                toolchain.name().to_string()
```
This may need tweaking if you agree with my above comment and decide that the channel may be optional in the toml.

kinnison commented on src/cli/rustup_mode.rs:
```diff
@@ -786,7 +786,7 @@ fn default_(cfg: &Cfg, m: &ArgMatches<'_>) -> Result<utils::ExitCode> {
 
         let cwd = utils::current_dir()?;
-        if let Some((toolchain, reason)) = cfg.find_override(&cwd)? {
+        if let Some((toolchain, _override_cfg, reason)) = cfg.find_override(&cwd)? {
```
Ditto, may need tweaking.

kinnison commented on src/cli/setup_mode.rs:
```diff
@@ -16,7 +16,7 @@ pub fn main() -> Result<utils::ExitCode> {
     }
 
-    // Internal testament dump used during CI.  Not for users.
+    // Internal testament dump used during CI. Not for users.
```
Whitespace bogon crept in, please get rid of this.

kinnison commented on src/config.rs:
```diff
@@ -18,6 +20,27 @@ use crate::settings::{Settings, SettingsFile, DEFAULT_METADATA_VERSION};
+
+#[derive(Debug, Default, Deserialize, PartialEq, Eq)]
+struct ToolchainSection {
+    channel: String,
```
My preference would be `channel: Option<String>` as mentioned above.

kinnison commented on src/config.rs:
```diff
@@ -37,6 +60,21 @@ impl Display for OverrideReason {
 }
 
+#[derive(Debug, Default)]
+pub struct OverrideCfg {
```
If we make channel optional as per the above, then this struct becomes less necessary perhaps?  Otherwise if you want to ensure that there remains API separation between the `rust-toolchain` format and the internal structs used, then this should probably contain the `channel: Option<Toolchain<'_>>` too.

ebroto replied:
I think decoupling them is a good idea as you mention. Also, I've realized that this struct does not need to be exposed outside the `config` module, so I kept the previous API of `find_override` returning the toolchain and the override reason.

kinnison replied:
Sounds good.

kinnison commented on src/config.rs:
```diff
@@ -510,26 +559,74 @@ impl Cfg {
+        if contents.lines().any(|line| line.starts_with("[toolchain]")) {
+            toml::from_str::<OverrideFile>(contents)
+                .map(|file| Some(file))
+                .map_err(|e| ErrorKind::ParsingOverride(e).into())
```
If `channel` is made optional, here is where you'd need to verify that the constructed `OverrideFile` has at least one of `channel`, `components`, or `targets` specified.

kinnison commented on tests/cli-rustup.rs:
```diff
@@ -1402,6 +1402,174 @@ fn file_override_with_archive() {
+        .unwrap();
+
+        expect_stdout_ok(config, &["rustc", "--version"], "hash-nightly-1");
+        expect_stdout_ok(
```
Ideally if `channel` were optional, this test would **a)** not change channel since it's purpose is to test that the component is added and **b)** ensure that the component you select is *NOT* installed before the toolchain file is written out.  (Note, this is kinda like the below test)

ebroto replied:
I completely agree with `b)`, I've changed all the related tests to check that :+1:. 

With respect to `a)`, in this particular case I was trying to test that if the toolchain file specifies a toolchain that is not installed it would be installed before adding the component. Let me know if I misunderstood how rustup works with respect to this specific case. 

I've renamed and reordered a bit the tests to try to make my intent more explicit. Also I took advantage of the optional channel to reduce some test cases.

kinnison replied:
It's fine that you want to verify the toolchain gets installed; in that context, simply don't install the other toolchain too.  Let's try and keep the test as fast as it can be.

kinnison commented on tests/cli-rustup.rs:
```diff
@@ -1402,6 +1402,174 @@ fn file_override_with_archive() {
+#[test]
+fn file_override_toml_format_add_components_toolchain_misses_components() {
+    setup(&|config| {
+        expect_ok(config, &["rustup", "default", "stable"]);
```
Why not `rustup default nightly-2015-01-01` rather than installing two toolchains in this test?

kinnison commented on tests/cli-rustup.rs:
```diff
@@ -1402,6 +1402,174 @@ fn file_override_with_archive() {
+        .unwrap();
+
+        expect_stdout_ok(config, &["rustc", "--version"], "hash-nightly-1");
+        expect_stdout_ok(
+            config,
+            &["rustup", "component", "list"],
+            "rust-src (installed)",
+        );
```
All this does is confirm `rust-src` is available *afterwards* not that it also was not available *before* the toolchain file was written to disk.  Pleaes verify that the component is not installed beforehand.  otherwise changes in default profiles etc, may cause this test to become useless in a way that does not cause a test failure.

kinnison commented on tests/cli-rustup.rs:
```diff
@@ -1402,6 +1402,174 @@ fn file_override_with_archive() {
+        .unwrap();
+
+        // The toolchain should be installed and the invalid component skipped
+        expect_stdout_ok(config, &["rustc", "--version"], "hash-nightly-1");
```
Is it worth using a variant of this test which looks for the message that `rust-bongo` is skipped?

kinnison commented on tests/cli-rustup.rs:
```diff
@@ -1402,6 +1402,174 @@ fn file_override_with_archive() {
+fn file_override_toml_format_add_components_invalid_component() {
+    setup(&|config| {
+        expect_ok(config, &["rustup", "default", "stable"]);
+        expect_ok(
```
Again, why installing two toolchains here?

kinnison commented on tests/cli-rustup.rs:
```diff
@@ -1402,6 +1402,174 @@ fn file_override_with_archive() {
+fn file_override_toml_format_add_components_toolchain_misses_target() {
+    setup(&|config| {
+        expect_ok(config, &["rustup", "default", "stable"]);
+        expect_ok(
```
Again, it's not really worth installing both toolchains here.

kinnison commented on tests/cli-rustup.rs:
```diff
@@ -1402,6 +1402,174 @@ fn file_override_with_archive() {
+            config,
+            &["rustup", "component", "list"],
+            "arm-linux-androideabi (installed)",
+        );
```
Again, please verify the target is not installed before you write the toolchain file out.

kinnison reviewed (commented):
Over-all I'm very impressed.  The single main concern I have is that I think it'd be beneficial to make `channel` optional so that people can use this new format to specify components/targets without constraining the channel of the toolchain.  With that, a bit of refactoring would be appropriate since the result of finding an override configuration might be that there's no toolchain override (though that might come from elsewhere) but there might be components/targets to install.

In addition, the integration tests could be tweaked to be more correct, less prone to risk of silently becoming worthless in the future, and possibly quicker to run, with the tweaks I mention in comments.

This is an excellent first pass.  What do you think about my 'optional `channel`' point?

ebroto committed:
Apply suggestions from PR review

- Make channel optional in the TOML toolchain file.
- Avoid exposing OverrideCfg outside the `config` module.
- Rework tests to avoid installing multiple toolchains unnecessarily and
  improve test naming to express better the intent.
- Restore unrelated extra whitespace.

ebroto commented:
I think making the channel optional is a great idea, thanks for the suggestion!

I've also tweaked the integration tests with your insights. I'll leave some comments to clarify some decisions I made.

kinnison reviewed (commented):
I think this is looking pretty good now.  @rbtcollins Do you have time to skim?

rbtcollins commented on README.md:
````diff
@@ -404,9 +404,21 @@ source repository. This is most often the case for nightly-only
+```
+
+If the TOML format is used, the `[toolchain]` section and the `channel` entry
+are mandatory.
````
This prose needs updating.

rbtcollins commented on src/config.rs:
```diff
@@ -18,6 +19,35 @@ use crate::settings::{Settings, SettingsFile, DEFAULT_METADATA_VERSION};
+    }
+}
+
+enum OverrideKind {
```
I'm not sure we want to be passing around an enum to the rest of the code: I think we just want to upgrade the old serialisation format to the new model, and always pass around an OverrideFile.

rbtcollins commented on src/config.rs:
```diff
@@ -510,26 +588,90 @@ impl Cfg {
+    fn parse_override_file<S: AsRef<str>>(contents: S) -> Result<Option<OverrideFile>> {
+        let contents = contents.as_ref();
+
+        if contents.lines().any(|line| line.starts_with("[toolchain]")) {
```
I highly recommend not taking this approach, which is fragile: toml permits sections in any order, and any future changes we make to the schema will make this code more and more awkward.

What I suggest instead is to attempt to deserialise using serde directly. Because the two branches are serialised differently, you'll need to deserialise it twice rather than using serdes https://serde.rs/enum-representations.html#untagged feature. 

So try once, first, using the TOML struct, and toml format, and the second entry the plain string using https://docs.rs/crate/serde_plain/0.3.0.

That way you're not manually implementing any part of the underlying plumbing, and we can change it further without changing the deserialiser.

Also, this should give us utf8 support rather than US-ASCII.

ebroto replied:
@rbtcollins @kinnison following our discussion on Discord, I've applied this suggestion by avoiding to manually parse the TOML format, and showing the TOML parse error in addition to the "unknown channel" error if parsing as legacy format yields an invalid toolchain name. I've not used serde-plain because (correct me if I'm wrong) it would not respect the old behavior given that lines following the first one were simply ignored. I believe that same reason prevents trying to parse as legacy format before TOML.

I find this solution is not ideal, but it works. There's still one problem though, we can't control the error message for old rustups, as the contents of the first line could be anything. We can't require the first line to be anything specific if we want to avoid parsing TOML manually (which I think is wise advice, as e.g. `[ toolchain ]` is valid TOML).

All in all, I strongly feel that having a specific `rustup.toml` would get rid of all these problems, if I may propose it again at this stage (one last time, I promise :)). We can prevent having both files by erroring out so diverging contents should not be a concern. The only drawback I can think of is that it could make the integration tests slower, which I believe is currently a concern.

Let me know your thoughts.

rbtcollins replied:
the failure mode though is that older rustups will then behave as though there is no rustup file at all, which will be mysterious.

kinnison replied:
I'd rather risk unusual but failing behaviour from an old `rustup` over silent miscompilation with the wrong toolchain, or unusual errors as a result.  Thus no, `rustup.toml` is not going to be a thing at this point.

The way I see it, a reasonable approach would be:

1. An empty `rust-toolchain` file is a specific error, perhaps "Empty `rust-toolchain` file detected.  Please remove it, or else specify the desired toolchain properties in the file".
2. A single-line `rust-toolchain` file is to be treated as the old-form, even if it's `[toolchain]` or similar, the resultant error will probably be obvious
3. Any other file content should be parsed-as-toml and errors reported as appropriate.

ebroto replied:
Alright, I totally understand the concerns about the old rustups, I'll implement the change as described.

Just to explain the way I understood it, is that pre-`rust-toolchain` rustups were ignoring the file too.

rbtcollins replied:
Fair enough, but that was 3 years ago (June 2017), and we can't do anything about those versions: if someone on RHEL or Ubuntu 1604 or whatnot running an LTS version of rustup tries to build, they will get unexpected results, and we may have to support them; but we are at least not creating new silent failure modes: any newer rustup will fail verbosely.

rbtcollins reviewed (commented):
A few thoughts

ebroto committed:
Remove `OverrideKind` enum

ebroto committed:
Avoid parsing TOML manually

When trying to detect the format of the toolchain override file,
let serde do all the work. If the file does not parse as TOML,
fallback to the legacy format, but keep the TOML error. If the first
line does not happen to be a valid toolchain, show also the TOML error
so that TOML syntax errors are reported.

ebroto committed:
Update README: channel not mandatory in override

rbtcollins commented on src/config.rs:
```diff
@@ -37,6 +73,26 @@ impl Display for OverrideReason {
+
+impl<'a> OverrideCfg<'a> {
+    fn from_file(cfg: &'a Cfg, file: OverrideFile) -> Result<Self> {
+        Ok(Self {
```
Does a result (None,None,None) make sense? what will happen when someone has such a config file - will it produce an error? I see the merit in permitting configs where at least one of the fields is not none, but with all none - do we want to treat that as an absent file, or as an error? @kinnison

ebroto replied:
I went for treating this case as an empty file in the legacy format, which was not yielding errors.

kinnison replied:
A file consisting of `[toolchain]` only (i.e. the `(None, None, None)` case) is an error IMO.  You must be specifying *something* to warrant a `rust-toolchain` file - similarly, an empty such file is worthless and thus an error as I mentioned in a previous comment.

kinnison replied:
(While previously an empty file was not an error, I think from now on it *should* be.  and I'm prepared to accept that as a breaking change if people view it as such.)

ebroto replied:
With the latest version `OverrideFile` is guaranteed to have at least one `Some`.

rbtcollins commented on src/config.rs:
```diff
@@ -510,26 +589,97 @@ impl Cfg {
+        let contents = contents.as_ref();
+
+        // Avoid TOML parsing error for backwards compatibility with the legacy format
+        if contents.is_empty() {
```
This could go in the 'in case of error' block I think, or does lines() not return an empty line for an empty file?

ebroto replied:
You are right, in fact, I think I can remove the code completely as long as we ignore the toml error

ebroto replied:
This is not necessary anymore with the new parsing strategy.

ebroto committed:
Rework toolchain file parsing
